### ISSUE TITLE
Replacing Badger with Pebble DB

### ISSUE THREAD

**Issue by j1010001** [onflow/flow-go#6515]
## Why (Objective)

1. Badger DB is a project that is not maintained anymore. At some point it will not be compatible with a new version of Go and will block the upgrade of flow-go to newer version of Go language.
2. Badger has caused memory spikes and performance issues in the past (so far mitigated with more aggressive GC GoMemLimit setting) and as the load on the DB grows these issues are likely to reoccur in the future.
3. Badger DB does not support data pruning, so on long-running network removing unused data requires node downtime.
4. Pebble supports data compression, so reduces disk growth rate

## Key Result:
Access, Verification and Execution nodes switch to using Pebble during the Q3 spork.

## Stretch goal:
Collection and Consensus switch to using Pebble during the Q3 spork.

## DACI

|Driver|Approver|Consulted|Informed|
|:-:|:-:|:-:|:-:|
| @zhangchiqing   | @j1010001    | @Kay-Zee , @AlexHentschel, Flow Protocol Engineering   | Flow Protocol Engineering   |

---
We have identified following project milestones  that complete the work of removing flow-go dependency on Badger DB.

## ✅ Milestone 1 - Refactor data access & prune chunk data packs

- [x] https://github.com/onflow/flow-go/issues/6516

## Milestone 2 - DB access refactoring for low-risk data on EN, VN and AN

- [x] https://github.com/onflow/flow-go/issues/6527

## Milestone 3 - unblock pruning of Execution, Access and Verification data

- [ ] https://github.com/onflow/flow-go/issues/7242

## Milestone 4 - DB access refactoring - remove dependency on Badger DB completely form ENs and ANs

- [x] https://github.com/onflow/flow-go/issues/7265
 
## Milestone 5 - DB access refactoring - remove dependency on Badger DB completely form Collection and Consensus nodes

OKR placeholder: #6528 

**Consensus**
- [x] Protocol Data (Consensus Builder)
    - [x] [BeaconKey](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L319), [DKG State](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L219)
    - [x] Consensus [Builder](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L735), [Finalizer](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L561), [Persister](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L606)
    - [x] [ExecForkSuppressor](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L383)
- [x] Deployment - Switching from Badger to Pebble (Dynamic bootstrap (preferred), spork, or Migration)
- [x] Protocol Data (Consensus Follower)

**Collection**
- [x] Collection Consensus
    - [x] [Cluster State](https://github.com/onflow/flow-go/blob/master/cmd/collection/main.go#L494)
    - [x] [Guarantee creation](https://github.com/onflow/flow-go/blob/master/storage/badger/procedure/cluster.go#L162
    - [x] [Collections Provider](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/collection/main.go#L452) 
- [x] Deployment - Switching from Badger to Pebble (Dynamic bootstrap (preferred), spork, or Migration)
    
- Tools
TBD

## Milestone 6 - pruning of Execution, access and verification data
Task breakdown TDB

## Milestone 7 - remaining tasks
- [ ] https://github.com/onflow/flow-go/issues/7682

- [x] https://github.com/onflow/flow-go/issues/7126

**Comment by fxamacker** [comment]
> Badger DB is a project that is not maintained anymore.

@j1010001 did you mean that BadgerDB **_v2_** and **_v3_** are not maintained anymore (instead of entire BadgerDB project)?

BadgerDB released v4.0 in Feb 2023 and v4.3 in Aug 2024.

- [v4.3.0](https://github.com/dgraph-io/badger/releases/tag/v4.3.0) (Aug 28, 2024)
- [v3.2103.5](https://github.com/dgraph-io/badger/releases/tag/v3.2103.5) (Dec 15, 2022) is the last v3.
- [v2.2007.4](https://github.com/dgraph-io/badger/releases/tag/v2.2007.4) (Aug 25, 2021) is the last v2 and version currently used by flow-go ([go.mod](https://github.com/onflow/flow-go/blob/master/go.mod)).

More details and other releases at [BadgerDB releases](https://github.com/dgraph-io/badger/releases).

**Comment by zhangchiqing** [comment]
Issues for this epic:

- https://github.com/onflow/flow-go/issues/6516
- https://github.com/onflow/flow-go/issues/6518
- https://github.com/onflow/flow-go/issues/6519
- https://github.com/onflow/flow-go/issues/6520
- https://github.com/onflow/flow-go/issues/6521
- https://github.com/onflow/flow-go/issues/6522
- https://github.com/onflow/flow-go/issues/6523

**Comment by j1010001** [comment]
> Issues for this epic:
> 
> * [Chunk Data pack Pruner #6516](https://github.com/onflow/flow-go/issues/6516)
> * [Replace Badger Transaction with Batch updates - Access Node-related data #6518](https://github.com/onflow/flow-go/issues/6518)
> * [Replace Badger Transaction with Batch updates - Verification Node-related data  #6519](https://github.com/onflow/flow-go/issues/6519)
> * [Replace Badger Transaction with Batch updates - Execution Node-related data  #6520](https://github.com/onflow/flow-go/issues/6520)
> * [Replace Badger Transaction with Batch updates - Collection Node-related data #6521](https://github.com/onflow/flow-go/issues/6521)
> * [Replace Badger Transaction with Batch updates - Protocol-related data  #6522](https://github.com/onflow/flow-go/issues/6522)
> * [Database Operation Transition: From Badger to Pebble #6523](https://github.com/onflow/flow-go/issues/6523)

All the issues above have been linked to the relevant phases

**Comment by bluesign** [comment]
did you have any preliminary results with running nodes with pebble instead of badger? I am using pebble and pretty happy so far, but was thinking to check badger for faster writes, though I am scared on memory usage etc a bit.

**Comment by zhangchiqing** [comment]
Unfortunately , we didn’t gather metrics for the proof-of-concept benchmark, as our focus was on ensuring execution correctness. Once [this issue](https://github.com/onflow/flow-go/issues/6527) is completed, I will collect metrics for comparison.

**Comment by zhangchiqing** [comment]
Trying to list all the data and their location that need to be refactored.

- Execution
    - Execution Result https://github.com/onflow/flow-go/pull/6906
    - Collections https://github.com/onflow/flow-go/pull/7059
    - [StopControl](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/execution_builder.go#L811) (VersionBeacon) 
    - [Execution Data](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/execution_builder.go#L679-L680) (Bitswap) 
    - [Result GCP Uploader](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/execution_builder.go#L446)
    - Switching from Badger to Pebble
        - Migrate [the last executed result](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/engine/execution/ingestion/core.go) to pebble in order to make next block executable
    - [Follower](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/execution_builder.go#L1146)

- Verification
    - Approvals https://github.com/onflow/flow-go/pull/6868
    - ChunkQueue https://github.com/onflow/flow-go/pull/6947
    - Switching from Badger to Pebble https://github.com/onflow/flow-go/pull/6948
    - [Follower](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/verification_builder.go#L335)

- Access
    - [Execution Data](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/access/node_builder/access_node_builder.go#L952) ([Bitswap](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/access/node_builder/access_node_builder.go#L859))
    - Execution Result https://github.com/onflow/flow-go/pull/6906
    - [Collections](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/access/node_builder/access_node_builder.go#L2119)
    - [StopControl](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/access/node_builder/access_node_builder.go#L1880) (VersionBeacon)
    - [Tx](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/access/node_builder/access_node_builder.go#L1849)/[Block](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/access/node_builder/access_node_builder.go#L2153) Status RPC API 
    - Switching from Badger to Pebble
    - [Follower](https://github.com/onflow/flow-go/blob/master/cmd/access/node_builder/access_node_builder.go#L378)

- Follower
    - Protocol Data (Consensus Follower)
    - [Libp2p Blocklist](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/scaffold.go#L1233)

- Consensus
    - Protocol Data (Consensus Builder)
        - [BeaconKey](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L319), [DKG State](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L219)
        - Consensus [Builder](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L735), [Finalizer](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L561), [Persister](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L606)
        - [ExecForkSuppressor](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L383)
    - Switching from Badger to Pebble

- Collection
    - Collection Consensus
        - [Cluster State](https://github.com/onflow/flow-go/blob/master/cmd/collection/main.go#L494)
        - [Guarantee creation]https://github.com/onflow/flow-go/blob/master/storage/badger/procedure/cluster.go#L162
    - [Collections Provider](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/collection/main.go#L452) 
    - Switching from Badger to Pebble
    - [Follower](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/collection/main.go#L331)

- Tools
    - Utility
        - add flags to [read](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/util/cmd/read-badger/cmd/root.go#L30) from pebble
        - Test if pebble can read data without stopping the process by [creating a checkpoint](https://github.com/cockroachdb/pebble/blob/master/checkpoint.go#L143) of db on the fly.
    - AdminTool
        - If Utility can read pebble without stopping the process, we can get rid of [the admin tool that reads data ](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/scaffold.go#L2000-L2006)from pebble.

**Comment by vishalchangrani** [comment]
hi @j1010001, @zhangchiqing - is my understanding correct w.r.t outcomes of the milestones above:

1. Milestone 1 - chunkdata pack pruning on EN ✅ 

2. Milestone 2 - low risk data on EN, AN, VN moved to Pebble DB (in progress)

3. Milestone 3 - 
     - Follower engine migrated to PebbleDB.
     - Verification node can run exclusively on PebbleDB.

4. Milestone 4 - EN and AN can run exclusively on PebbleDB

5. Milestone 5- SN and LN can run exclusively on PebbleDB

6. Milestone 6 - Data can be automatically pruned on EN, AN, VN.

7. Milestone 7 - Upgrade to Pebble 2.x

**Comment by j1010001** [comment]
Scope update:

There are portions of M5 Consensus that are moving to M3:

Consensus [Builder](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L735), [Finalizer](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L561), [Persister](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L606)

[ExecForkSuppressor](https://github.com/onflow/flow-go/blob/21b3a7bbaec72f238b4a893e91fa06505c3a4e98/cmd/consensus/main.go#L383)

Deployment - Switching from Badger to Pebble (Dynamic bootstrap (preferred), spork, or Migration)

Protocol Data (Consensus Follower)

**Comment by j1010001** [comment]
Removed M7 - upgrade to Pebble 2.x (moved as task to M4)

**Comment by vishalchangrani** [comment]
- One node of each type is now running on Pebble db on mainnet (Consensus node still has some DKG data stored in badgerdb, see [issue](https://github.com/onflow/flow-go/issues/7695)).
- Alex is back and he is working through Leo’s PR. There are 4 of these which are on the critical path.
- The next big milestone for this OKR is merging Malleability into PebbleDB, resolving conflicts and then testing the combine branch. The plan is to merge malleability to master by this week, then work through merging malleability to pebble DB next two weeks. Rough ETA is end of august.
